### ISSUE TITLE
Add method for `nothing .= `

### ISSUE THREAD

**Issue by Sbozzolo** [CliMA/ClimaCore.jl#1605]
A common pattern that is seen everywhere in the ClimaAtmos diagnostics is something this:
```julia
        if isnothing(out)
            return TD.dry_pottemp.(thermo_params, cache.precomputed.ᶜts)
        else
            out .= TD.dry_pottemp.(thermo_params, cache.precomputed.ᶜts)
        end
```
What we are trying to do is this: when `out` is `nothing`, we want to allocate a new `Field` and return it. When `out` is a preallocated `Field`, we just want to use it. It would be really nice for ClimaCore to support this in a single `out .= something` expression.

**Comment by charleskawczynski** [comment]
I don't know that we want to overload `nothing .=`, but I do agree that it would be convenient to have a lazy and eager way to compute a broadcast expression (this would be helpful for unit tests and performance, too). Mind if we update the title?

**Comment by Sbozzolo** [comment]
> I don't know that we want to overload `nothing .=`, but I do agree that it would be convenient to have a lazy and eager way to compute a broadcast expression (this would be helpful for unit tests and performance, too). Mind if we update the title?

Please, go ahead! I don't have any concrete idea on how to accomplish this.

**Comment by dennisYatunin** [comment]
Yes, I'm currently in the process of sketching out how we could add something like `@lazydot` and/or `@fuseddot` to ClimaCore in order to simplify and optimize our broadcasting. Ideally, this will allow us to fully separate code related to broadcast performance/allocations from code related to physics (i.e., the right hand sides of equations).

**Comment by charleskawczynski** [comment]
xref: https://github.com/MasonProtter/DontMaterialize.jl
